org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/internal/telemetry/SonarLintTelemetry.java

```diff
@@ -0,0 +1,172 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarlint.eclipse.core.internal.telemetry;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.MalformedURLException;
+import java.net.Proxy;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.util.concurrent.TimeUnit;
+import org.eclipse.core.net.proxy.IProxyData;
+import org.eclipse.core.net.proxy.IProxyService;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
+import org.sonarlint.eclipse.core.SonarLintLogger;
+import org.sonarlint.eclipse.core.internal.SonarLintCorePlugin;
+import org.sonarlint.eclipse.core.internal.event.AnalysisEvent;
+import org.sonarlint.eclipse.core.internal.event.AnalysisListener;
+import org.sonarlint.eclipse.core.internal.resources.ProjectsProviderUtils;
+import org.sonarlint.eclipse.core.internal.utils.SonarLintUtils;
+import org.sonarsource.sonarlint.core.client.api.common.TelemetryClientConfig;
+import org.sonarsource.sonarlint.core.telemetry.Telemetry;
+
+public class SonarLintTelemetry implements AnalysisListener {
+  private static final String PRODUCT = "SonarLint Eclipse";
+  private static final String STORAGE_FILENAME = "sonarlint_usage";
+  public static final String DISABLE_PROPERTY_KEY = "sonarlint.telemetry.disabled";
+  private boolean enabled;
+  private Telemetry telemetryEngine;
+
+  private TelemetryJob scheduledJob;
+
+  private static Path getStorageFilePath() {
+    return SonarLintCorePlugin.getDefault().getStateLocation().toFile().toPath().resolve(STORAGE_FILENAME);
+  }
+
+  public void optOut(boolean optOut) {
+    if (telemetryEngine != null) {
+      if (optOut == !telemetryEngine.enabled()) {
+        return;
+      }
+      telemetryEngine.enable(!optOut);
+      if (optOut) {
+        try {
+          TelemetryClientConfig clientConfig = getTelemetryClientConfig();
+          telemetryEngine.getClient().optOut(clientConfig, isAnyProjectConnected());
+        } catch (Exception e) {
+          // Silently ignore
+        }
+      }
+    }
+  }
+
+  public boolean enabled() {
+    return enabled;
+  }
+
+  public boolean optedIn() {
+    return enabled && this.telemetryEngine.enabled();
+  }
+
+  public void init() {
+    if ("true".equals(System.getProperty(DISABLE_PROPERTY_KEY))) {
+      this.enabled = false;
+      SonarLintLogger.get().info("Telemetry disabled by system property");
+      return;
+    }
+    try {
+      this.telemetryEngine = new Telemetry(getStorageFilePath(), PRODUCT, SonarLintUtils.getPluginVersion());
+      SonarLintCorePlugin.getAnalysisListenerManager().addListener(this);
+      this.scheduledJob = new TelemetryJob();
+      scheduledJob.schedule(TimeUnit.MINUTES.toMillis(1));
+      this.enabled = true;
+    } catch (Exception e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)